dev/tools/roll_dev.dart

```diff
@@ -0,0 +1,150 @@
+    print('will delete files! Run with -n to find out which ones.');
+    exit(1);
+  }
+
+  String version = getFullTag();
+  final Match match = parseFullTag(version);
+  if (match == null) {
+    print('Could not determine the version for this build.');
+    if (version.isNotEmpty)
+      print('Git reported the latest version as "$version", which does not fit the expected pattern.');
+    exit(1);
+  }
+
+  final List<int> parts = match.groups(<int>[1, 2, 3]).map(int.parse).toList();
+
+  if (match.group(4) == '0') {
+    print('This commit has already been released, as version ${parts.join(".")}.');
+    exit(0);
+  }
+
+  switch (level) {
+    case kX:
+      parts[0] += 1;
+      break;
+    case kY:
+      parts[1] += 1;
+      break;
+    case kZ:
+      parts[2] += 1;
+      break;
+  }
+  version = parts.join('.');
+
+  runGit('fetch upstream', 'fetch upstream');
+  runGit('reset upstream/master --hard', 'check out master branch');
+  runGit('tag $version', 'tag the commit with the version label');
+
+  print('Your tree is ready to publish Flutter $version to the "dev" channel.');
+  stdout.write('Are you? [yes/no] ');
+  if (stdin.readLineSync() != 'yes') {
+    runGit('tag -d $version', 'remove the tag you did not want to publish');
+    print('The dev roll has been aborted.');
+    exit(0);
+  }
+
+  runGit('push upstream $version', 'publish the version');
+  runGit('push upstream HEAD:dev', 'land the new version on the "dev" branch');
+  print('Flutter version $version has been rolled to the "dev" channel!');
+}
+
+String getFullTag() {
+  return getGitOutput(
+    'describe --match v*.*.* --exclude *-* --first-parent --long --tags',
+    'obtain last released version number',
+  );
+}
+
+Match parseFullTag(String version) {
+  final RegExp versionPattern = new RegExp('^v([0-9]+)\.([0-9]+)\.([0-9]+\)-([0-9]+)-g([a-f0-9]+)\$');
+  return versionPattern.matchAsPrefix(version);
+}
+
+String getGitOutput(String command, String explanation) {
+  final ProcessResult result = Process.runSync('git', command.split(' '));
+  if (result.stderr.isEmpty && result.exitCode == 0)
+    return result.stdout.trim();
+  if (result.exitCode != 0) {
+    print('Failed to $explanation. Git exitted with error code ${result.exitCode}.');
+  } else {
+    print('Failed to $explanation.');
+  }
+  if (result.stdout.isNotEmpty)
+    print('stdout from git:\n${result.stdout}\n');
+  if (result.stderr.isNotEmpty)
+    print('stderr from git:\n${result.stderr}\n');
+  exit(1);
+  return null; // for the analyzer's sake
+}
+
+void runGit(String command, String explanation) {
+  final ProcessResult result = Process.runSync('git', command.split(' '));
+  if (result.exitCode != 0) {
+    print('Failed to $explanation. Git exitted with error code ${result.exitCode}.');
+    if (result.stdout.isNotEmpty)
+      print('stdout from git:\n${result.stdout}\n');
+    if (result.stderr.isNotEmpty)
+      print('stderr from git:\n${result.stderr}\n');
+    exit(1);
```

Comment:
result.exitCode instead of `1`?

Reply:
In general I'm not a fan of propagating exit codes that way, because it's essentially merging two namespaces (theirs and ours and any other processes we might do this to). In practice what matters is 0 vs 1.